### PR TITLE
Feature/negamax

### LINKED ISSUE #47: [Rust] Implémenter un mini calcul heuristique préliminaire
Afin de pouvoir déterminer quel node mérite d'être calculé en profondeur il faudrait créer une méthode sur goban d'évaluation rapide de l'heuristique.
A définir plus en profondeur.

### LINKED ISSUE #53: Implement NegaMax algorithm
We should get some performance gains from implementing this feature, it should not be hard to do since negamax is a variant of minimax, so copying and pasting then modifying it should work just fine.
Should be implemented in a new file, named negamax.rs, at the same level as minimax.rs

### PR DESCRIPTION
**Pull request by Ragarnoy**
Solves #53 and #47, performance gain is disappointing, need to close a bunch of branches after this because it was a fucking mess

- [x] im tired

Merge Second

### CODE REVIEW

File: src/algorithm/mod.rs
Comment (Ragarnoy) [inline src/algorithm/mod.rs:83]: not sure about this one might be a rebase fuckup

File: src/algorithm/mod.rs
Comment (Ragarnoy) [inline src/algorithm/mod.rs:87]: same

File: src/node/mod.rs
Comment (arlaine4) [inline src/node/mod.rs:177]: grave plus lisible 👍🏼

File: src/goban/mod.rs
Comment (arlaine4) [inline src/goban/mod.rs]: maybe rename this func what do you think ? Not very exhaustive method name

**Review by arlaine4 — APPROVED** [review]

File: src/algorithm/mod.rs
Comment (demonoidv) [reply src/algorithm/mod.rs:83]: But you merged the PR anyway, this is not the way it should work.
And yes indeed, there probably was an issue with a rebase of something like that. I know it's not the fun part of the job but we can't afford the lux to mess with something as important as the heuristic function itself.

File: src/bitboard/pattern.rs
Comment (demonoidv) [inline src/bitboard/pattern.rs:524]: Why?

File: src/bitboard/pattern.rs
Comment (demonoidv) [inline src/bitboard/pattern.rs:555]: Why?

File: src/bitboard/pattern.rs
Comment (demonoidv) [inline src/bitboard/pattern.rs:578]: Why?

File: src/lib.rs
Comment (demonoidv) [inline src/lib.rs:88]: As already said in this review, we can't let that kind of change pass, it literally breaks the AI.

File: src/lib.rs
Comment (demonoidv) [inline src/lib.rs:103]: How is it even possible for this to happen?

I don't want to play the cop here but @tuturr, you took the responsibility to review (and approve, it really means something serious, don't approve any PR unless you are **absolutely** sure that everything is OK), let's say it's ok for this one as you are still new at this but you should not have approved this PR, because you should have performed a real code review and saw this line duplication. Yes, it's boring, nobody wants to do a code review, but if nobody actually does I can assure you that our project is really far from being functional.

For the next time, you have to know that by approving a PR, you actually say that you are taking part of the responsibility of the said PR. In other words, if something in a PR you approved breaks the product then you are as blamable as the author of the PR.

For anyone reading this comment (cc @Ragarnoy, @abassibe), as a reminder, doing code review is a real job, not something we do waiting for a build to complete and thinking about our next meal. We have (I included of course) to be dedicated to the task and really careful because it's the very last safety we have before considering the code as done and working properly.

**Review by demonoidv — COMMENTED** [review]
A bit late for the review, I know, and I apologize for that.
This PR brings some bugs, don't worry, I will take care of fixing those.

File: src/lib.rs
Comment (demonoidv) [reply src/lib.rs:88]: BTW, this is already fixed by [this commit](https://github.com/abassibe/Gomoku/commit/88c591ea696f1f4222cd4f592ec3618ac74f2b31).

**Pull request merged**
